pkg/pushsync/pushsync.go

```diff
@@ -219,80 +224,187 @@ func (ps *PushSync) PushChunkToClosest(ctx context.Context, ch swarm.Chunk) (*Re
+
+		// if you manage to get a tag, just increment the respective counter
+		t, err := ps.tagg.Get(ch.TagID())
+		if err == nil && t != nil {
+			err = t.Inc(tags.StateSent)
+			if err != nil {
+				return nil, err
+			}
+		}
 
-	//  if you manage to get a tag, just increment the respective counter
-	t, err := ps.tagg.Get(ch.TagID())
-	if err == nil && t != nil {
-		err = t.Inc(tags.StateSent)
+		// Check if the receipt is valid
+		if !ch.Address().Equal(swarm.NewAddress(receipt.Address)) {
+			ps.metrics.InvalidReceiptReceived.Inc()
+			_ = streamer.Reset()
+			return nil, fmt.Errorf("invalid receipt. peer %s", peer.String())
+		}
+
+		err = ps.accounting.Credit(peer, receiptPrice)
 		if err != nil {
 			return nil, err
 		}
-	}
 
-	receiptRTTTimer := time.Now()
-	receipt, err := ps.receiveReceipt(ctx, r)
-	if err != nil {
-		_ = streamer.Reset()
-		return nil, fmt.Errorf("receive receipt from peer %s: %w", peer.String(), err)
+		rec := &Receipt{
+			Address: swarm.NewAddress(receipt.Address),
+		}
+
+		return rec, nil
 	}
-	ps.metrics.ReceiptRTT.Observe(time.Since(receiptRTTTimer).Seconds())
 
-	// Check if the receipt is valid
-	if !ch.Address().Equal(swarm.NewAddress(receipt.Address)) {
-		ps.metrics.InvalidReceiptReceived.Inc()
-		_ = streamer.Reset()
-		return nil, fmt.Errorf("invalid receipt. peer %s", peer.String())
+	ps.logger.Tracef("pushsync-push: failed to push chunk %s: reached max peers of %v", ch.Address(), maxPeers)
+
+	if lastErr != nil {
+		return nil, lastErr
 	}
 
-	err = ps.accounting.Credit(peer, receiptPrice)
+	return nil, topology.ErrNotFound
+}
+
+// closestPeer returns address of the peer that is closest to the chunk with
+// provided address addr. This function will ignore peers with addresses
+// provided in skipPeers.
+func (ps *PushSync) closestPeer(addr swarm.Address, skipPeers []swarm.Address) (swarm.Address, error) {
+	closest := swarm.Address{}
+	err := ps.peerSuggester.EachPeerRev(func(peer swarm.Address, po uint8) (bool, bool, error) {
+		for _, a := range skipPeers {
+			if a.Equal(peer) {
+				return false, false, nil
+			}
+		}
+		if closest.IsZero() {
+			closest = peer
+			return false, false, nil
+		}
+		dcmp, err := swarm.DistanceCmp(addr.Bytes(), closest.Bytes(), peer.Bytes())
+		if err != nil {
+			return false, false, fmt.Errorf("distance compare error. addr %s closest %s peer %s: %w", addr.String(), closest.String(), peer.String(), err)
+		}
+		switch dcmp {
+		case 0:
+			// do nothing
+		case -1:
+			// current peer is closer
+			closest = peer
+		case 1:
+			// closest is already closer to chunk
+			// do nothing
+		}
+		return false, false, nil
+	})
 	if err != nil {
-		return nil, err
+		return swarm.Address{}, err
 	}
 
-	rec := &Receipt{
-		Address: swarm.NewAddress(receipt.Address),
+	// check if found
+	if closest.IsZero() {
+		return swarm.Address{}, topology.ErrNotFound
 	}
 
-	return rec, nil
+	return closest, nil
+}
+
+func (ps *PushSync) blocklistPeer(peer swarm.Address) {
```

Comment:
this should be abstracted in blocklist pkg

Reply:
Not sure about this, it's just separate method as its called on two places, and has specific logging depending if error or not.

Reply:
if such is needed in one protocol then inanother. one too